File: src/languages/python/binary/circuits.py
from converter import *
from config import *
from gates import *

def adder(a, b, carryIn):
    bina = DecompileString(FixBinary(a))
    binb = DecompileString(FixBinary(b))

    sum=[]
    e=BitLimit-1
    carry=carryIn
    for _ in range(BitLimit):
        sum1 = Xor(bina[e], binb[e])
        sum2 = Xor(sum1, carry)
        carry = Or(And(bina[e], binb[e]), And(sum1, carry))
        sum = sum+[sum2]
        e=e-1

    return CompileString(ReverseList(sum))

def subtractor(a, b):
    bina = DecompileString(FixBinary(a))
    binb = DecompileString(FixBinary(b))

    sum=[]
    e=BitLimit-1
    carry=0
    for _ in range(BitLimit):
        sum1 = Xor(Not(bina[e]), binb[e])
        sum2 = Xor(sum1, carry)
        carry = Or(And(Not(bina[e]), binb[e]), And(sum1, carry))
        sum = sum+[Not(sum2)]
        e=e-1

    return CompileString(ReverseList(sum))

def multiplier(a, b):
    exit()

print(subtractor("1000", "1100"))

File: src/languages/cs/binary/circuits.cs
namespace MinicallibCircuits{

    using MinicallibConfig;
    using MinicallibConverter;
    using MinicallibGates;

    public class circuits{

        public static string adder(string a, string b, bool carryIn){
            bool[] bina = converter.Decompilestring(converter.FixBinary(a));
            bool[] binb = converter.Decompilestring(converter.FixBinary(b));
        
            bool[] sum = new bool[config.BitLimit];
            int e = config.BitLimit-1;
            bool carry=carryIn;
            for (int i=0; i<config.BitLimit; i++) {
                bool sum1 = gates.Xor(bina[e], binb[e]);
                bool sum2 = gates.Xor(sum1, carry);
                carry = gates.Or(gates.And(bina[e], binb[e]), gates.And(sum1, carry));
                sum[i]=sum2;
                e--;
            }
        
            return converter.Compilestring(converter.ReverseList(sum));
        }
        
        public static string subtractor(string a, string b){
            bool[] bina = converter.Decompilestring(converter.FixBinary(a));
            bool[] binb = converter.Decompilestring(converter.FixBinary(b));
        
            bool[] sum = new bool[config.BitLimit];
            int e = config.BitLimit-1;
            bool carry=false;
            for (int i=0; i<config.BitLimit; i++) {
                var sum1 = gates.Xor(gates.Not(bina[e]), binb[e]);
                var sum2 = gates.Xor(sum1, carry);
                carry = gates.Or(gates.And(gates.Not(bina[e]), binb[e]), gates.And(sum1, carry));
                sum[i]=gates.Not(sum2);
                e--;
            }
        
            return converter.Compilestring(converter.ReverseList(sum));
        }

    }
}